src/lib/blkalloc/bitmap_blk_allocator.cpp

```diff
@@ -0,0 +1,201 @@
+        meta_service().register_handler(
+            get_name(),
+            [this](meta_blk* mblk, sisl::byte_view buf, size_t size) {
+                on_meta_blk_found(voidptr_cast(mblk), std::move(buf), size);
+            },
+            nullptr);
+    }
+
+    if (is_fresh || !is_persistent()) {
+        m_disk_bm = std::make_unique< sisl::Bitset >(m_num_blks, m_chunk_id, m_align_size);
+        do_init();
+    }
+}
+
+void BitmapBlkAllocator::do_init() {
+    // NOTE:  Blocks per portion must be modulo word size so locks do not fall on same word
+    m_blks_per_portion = sisl::round_up(m_blks_per_portion, m_disk_bm ? m_disk_bm->word_size() : 64u);
+
+    m_blk_portions = std::make_unique< BlkAllocPortion[] >(get_num_portions());
+    for (blk_num_t index{0}; index < get_num_portions(); ++index) {
+        m_blk_portions[index].set_portion_num(index);
+    }
+}
+
+void BitmapBlkAllocator::on_meta_blk_found(void* mblk_cookie, sisl::byte_view const& buf, size_t size) {
+    m_meta_blk_cookie = mblk_cookie;
+
+    m_disk_bm = std::unique_ptr< sisl::Bitset >{new sisl::Bitset{
+        hs_utils::extract_byte_array(buf, meta_service().is_aligned_buf_needed(size), meta_service().align_size())}};
+
+    m_alloced_blk_count.store(m_disk_bm->get_set_count(), std::memory_order_relaxed);
+    do_init();
+
+    load();
+}
+
+void BitmapBlkAllocator::cp_flush(CP*) {
+    if (m_is_disk_bm_dirty.load()) {
+        sisl::byte_array bitmap_buf = acquire_underlying_buffer();
+        if (m_meta_blk_cookie) {
+            meta_service().update_sub_sb(bitmap_buf->bytes, bitmap_buf->size, m_meta_blk_cookie);
+        } else {
+            meta_service().add_sub_sb(get_name(), bitmap_buf->bytes, bitmap_buf->size, m_meta_blk_cookie);
+        }
+        m_is_disk_bm_dirty.store(false); // No longer dirty now, needs to be set before releasing the buffer
+        release_underlying_buffer();
+    }
+}
+
+bool BitmapBlkAllocator::is_blk_alloced_on_disk(const BlkId& b, bool use_lock) const {
+    // for non-persistent bitmap nothing to compare. So always return true
+    if (!is_persistent()) { return true; }
+
+    if (use_lock) {
+        const BlkAllocPortion& portion = blknum_to_portion_const(b.blk_num());
+        auto lock{portion.portion_auto_lock()};
+        return m_disk_bm->is_bits_set(b.blk_num(), b.blk_count());
+    } else {
+        return m_disk_bm->is_bits_set(b.blk_num(), b.blk_count());
+    }
+}
+
+BlkAllocStatus BitmapBlkAllocator::alloc_on_disk(BlkId const& bid) {
+    if (!is_persistent()) { return BlkAllocStatus::FAILED; }
+
+    rcu_read_lock();
+    auto list = get_alloc_blk_list();
+    if (list) {
+        // cp has started, accumulating to the list
+        list->push_back(bid);
+    } else {
+        auto set_on_disk_bm = [this](auto& b) {
+            BlkAllocPortion& portion = blknum_to_portion(b.blk_num());
+            {
+                auto lock{portion.portion_auto_lock()};
+                if (!hs()->is_initializing()) {
```

Comment:
Can we also put a comment here by saying, during recovery, log replay might do alloc

Reply:
Ok, I think there is a comment in free_blks, will add here as well.